teamcity: failed tests on master: testrace/TestServer/version/6, testrace/TestServer/version/3, testrace/TestServer, testrace/TestServer/version/4, testrace/TestServer/version

The following tests appear to have failed:

[#332416](https://teamcity.cockroachdb.com/viewLog.html?buildId=332416):

```
--- FAIL: testrace/TestServer/version/6 (0.000s)
Test ended in panic.

------- Stdout: -------
W170823 16:10:48.312944 5048 sql/distsqlrun/server.go:220  [n1] version mismatch in flow request: 6; this node accepts 4 through 5
--- FAIL: testrace/TestServer/version/3 (0.000s)
Test ended in panic.

------- Stdout: -------
W170823 16:10:48.317221 5050 sql/distsqlrun/server.go:220  [n1] version mismatch in flow request: 3; this node accepts 4 through 5
--- FAIL: testrace/TestServer (0.000s)
Test ended in panic.

------- Stdout: -------
W170823 16:10:46.423989 4040 server/status/runtime.go:109  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I170823 16:10:46.436786 4040 server/config.go:534  [n?] 1 storage engine initialized
I170823 16:10:46.436863 4040 server/config.go:536  [n?] RocksDB cache size: 512 MiB
I170823 16:10:46.436907 4040 server/config.go:536  [n?] store 0: in-memory, size 100 MiB
I170823 16:10:46.496063 4040 server/node.go:345  [n?] **** cluster 49107cf0-1add-4c94-9f2c-ab1edb04f6f4 has been created
I170823 16:10:46.496187 4040 server/server.go:788  [n?] **** add additional nodes by specifying --join=127.0.0.1:44591
I170823 16:10:46.547595 4040 storage/store.go:1200  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170823 16:10:46.547839 4040 server/node.go:462  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=3.2 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=3266.00 p25=3266.00 p50=3266.00 p75=3266.00 p90=3266.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00}
I170823 16:10:46.547997 4040 server/node.go:327  [n1] node ID 1 initialized
I170823 16:10:46.548212 4040 gossip/gossip.go:311  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:44591" > attrs:<> locality:<> ServerVersion:<major_val:1 minor_val:0 patch:0 unstable:3 > 
I170823 16:10:46.548852 4040 storage/stores.go:303  [n1] read 0 node addresses from persistent storage
I170823 16:10:46.549099 4040 server/node.go:607  [n1] connecting to gossip network to verify cluster ID...
I170823 16:10:46.550422 4040 server/node.go:632  [n1] node connected via gossip and verified as part of cluster "49107cf0-1add-4c94-9f2c-ab1edb04f6f4"
I170823 16:10:46.550759 4040 server/node.go:404  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I170823 16:10:46.552828 4040 sql/executor.go:383  [n1] creating distSQLPlanner with address {tcp 127.0.0.1:44591}
I170823 16:10:46.597523 4040 server/server.go:880  [n1] starting https server at 127.0.0.1:41801
I170823 16:10:46.597630 4040 server/server.go:881  [n1] starting grpc/postgres server at 127.0.0.1:44591
I170823 16:10:46.597673 4040 server/server.go:882  [n1] advertising CockroachDB node at 127.0.0.1:44591
W170823 16:10:46.611531 4040 sql/jobs/registry.go:156  [n1] unable to get node liveness: node not in the liveness table
I170823 16:10:46.640504 4079 storage/replica_command.go:2703  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
E170823 16:10:46.744592 4252 storage/replica_proposal.go:538  [n1,s1,r1/1:/M{in-ax}] could not load SystemConfig span: must retry later due to intent on SystemConfigSpan
I170823 16:10:46.763964 4079 storage/replica_command.go:2703  [split,n1,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
I170823 16:10:46.798732 4040 sql/event_log.go:102  [n1] Event: "alter_table", target: 12, info: {TableName:eventlog Statement:ALTER TABLE system.eventlog ALTER COLUMN "uniqueID" SET DEFAULT uuid_v4() User:node MutationID:0 CascadeDroppedViews:[]}
I170823 16:10:46.850040 4079 storage/replica_command.go:2703  [split,n1,s1,r3/1:/{System/NodeL���-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I170823 16:10:46.873122 4040 sql/lease.go:341  [n1] publish: descID=12 (eventlog) version=2 mtime=2017-08-23 16:10:46.872141181 +0000 UTC
I170823 16:10:46.900202 4079 storage/replica_command.go:2703  [split,n1,s1,r4/1:/{System/NodeL���-Max}] initiating a split of this range at key /System/tsd [r5]
I170823 16:10:47.010156 4079 storage/replica_command.go:2703  [split,n1,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I170823 16:10:47.013714 4040 sql/event_log.go:102  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:node}
I170823 16:10:47.067215 4079 storage/replica_command.go:2703  [split,n1,s1,r6/1:/{System/tse-Max}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r7]
E170823 16:10:47.138698 4275 storage/replica_proposal.go:538  [n1,s1,r6/1:/{System/tse-Max}] could not load SystemConfig span: must retry later due to intent on SystemConfigSpan
I170823 16:10:47.158219 4079 storage/replica_command.go:2703  [split,n1,s1,r7/1:/{Table/System���-Max}] initiating a split of this range at key /Table/11 [r8]
I170823 16:10:47.219397 4040 sql/event_log.go:102  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:
                    User:node}
I170823 16:10:47.225580 4040 server/server.go:1024  [n1] done ensuring all necessary migrations have run
I170823 16:10:47.225665 4040 server/server.go:1026  [n1] serving sql connections
I170823 16:10:47.233975 4079 storage/replica_command.go:2703  [split,n1,s1,r8/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r9]
I170823 16:10:47.254146 4691 sql/event_log.go:102  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:44591} Attrs: Locality: ServerVersion:1.0-3} ClusterID:49107cf0-1add-4c94-9f2c-ab1edb04f6f4 StartedAt:1503504646550477949 LastUp:1503504646550477949}
I170823 16:10:47.329918 4079 storage/replica_command.go:2703  [split,n1,s1,r9/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r10]
I170823 16:10:47.446258 4079 storage/replica_command.go:2703  [split,n1,s1,r10/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r11]
I170823 16:10:47.513701 4079 storage/replica_command.go:2703  [split,n1,s1,r11/1:/{Table/14-Max}] initiating a split of this range at key /Table/15 [r12]
I170823 16:10:47.552159 4079 storage/replica_command.go:2703  [split,n1,s1,r12/1:/{Table/15-Max}] initiating a split of this range at key /Table/16 [r13]
I170823 16:10:47.682307 4079 storage/replica_command.go:2703  [split,n1,s1,r13/1:/{Table/16-Max}] initiating a split of this range at key /Table/17 [r14]
I170823 16:10:47.734652 4079 storage/replica_command.go:2703  [split,n1,s1,r14/1:/{Table/17-Max}] initiating a split of this range at key /Table/18 [r15]
I170823 16:10:47.774467 4079 storage/replica_command.go:2703  [split,n1,s1,r15/1:/{Table/18-Max}] initiating a split of this range at key /Table/19 [r16]
I170823 16:10:48.126426 4720 sql/event_log.go:102  [client=127.0.0.1:51662,user=root,n1] Event: "create_database", target: 50, info: {DatabaseName:test Statement:CREATE DATABASE test User:root}
I170823 16:10:48.129662 4079 storage/replica_command.go:2703  [split,n1,s1,r16/1:/{Table/19-Max}] initiating a split of this range at key /Table/50 [r17]
I170823 16:10:48.232191 4079 storage/replica_command.go:2703  [split,n1,s1,r17/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r18]
I170823 16:10:48.243768 4720 sql/event_log.go:102  [client=127.0.0.1:51662,user=root,n1] Event: "create_table", target: 51, info: {TableName:test.t Statement:CREATE TABLE test.t (a INT PRIMARY KEY, b INT) User:root}
--- FAIL: testrace/TestServer/version/4 (0.000s)
Race detected!

------- Stdout: -------
I170823 16:10:48.323304 4040 util/stop/stopper.go:475  quiescing; tasks left:
1      storage.intentResolver: processing intents
1      node.Node: batch
1      distsqlrun.ServerImpl: sync flow
1      [async] drain
I170823 16:10:48.323544 4040 util/stop/stopper.go:475  quiescing; tasks left:
1      storage.intentResolver: processing intents
1      distsqlrun.ServerImpl: sync flow
1      [async] drain
E170823 16:10:48.324172 5037 sql/distsqlrun/tablereader.go:198  [n1,TableReader=51] scan error: node unavailable; try another peer
I170823 16:10:48.324543 4040 util/stop/stopper.go:475  quiescing; tasks left:
1      storage.intentResolver: processing intents
1      distsqlrun.ServerImpl: sync flow
==================
Read at 0x00c421b3a4e0 by goroutine 334:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:128 +0x0
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*RowChannel).Push()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:320 +0x23b
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*outbox).Push()
      <autogenerated>:225 +0xa1
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).cancel()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:459 +0x35f
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Wait()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:387 +0x143
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).RunSyncFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:347 +0x114
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x104
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).RunSyncFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:349 +0x313
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun._DistSQL_RunSyncFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/api.pb.go:323 +0xd5
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processStreamingRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:891 +0x1185
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:985 +0x14a2
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:551 +0xb6

Previous write at 0x00c421b3a4e0 by goroutine 337:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:320 +0x0
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*outbox).ProducerDone()
      <autogenerated>:226 +0x6b
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).Close()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:316 +0x53
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*tableReader).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/tablereader.go:200 +0x428

Goroutine 334 (running) created at:
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:552 +0xb8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:333 +0x12f6
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:400 +0xb56
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:559 +0x1e7
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveHTTP2Transport()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:540 +0x6f8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:507 +0x5f9

Goroutine 337 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:371 +0x6b0
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).RunSyncFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:346 +0x106
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x104
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).RunSyncFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:349 +0x313
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun._DistSQL_RunSyncFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/api.pb.go:323 +0xd5
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processStreamingRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:891 +0x1185
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:985 +0x14a2
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:551 +0xb6
==================
--- FAIL: testrace/TestServer/version (0.000s)
Test ended in panic.


```
Please assign, take a look and update the issue accordingly.

--- a/docs/RFCS/query_cancellation.md
+++ b/docs/RFCS/query_cancellation.md
@@ -208,7 +208,7 @@ to any other nodes are also cancelled (by grpc).
 - Producer processor to consumer: This is a special case only for those processors that
 do a lot of processing before emitting any rows (such as the sorter). These processors will check the
 local flow context for cancellation, and return an error to their consumer if it gets cancelled.
-- syncFlowConsumer special case: Since `syncFlowConsumer` does not have any streams that
+- syncFlowConsumer special case: Since `syncFlowConsumer` on the gateway node does not have any streams that
 cross node boundaries or call `FlowStream`, an error will be manually pushed to it
 upon a cancellation request on the gateway node, marking it as closed to all of its producers.
 

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -15,10 +15,11 @@
 package sql
 
 import (
-	"golang.org/x/net/context"
+	"sync/atomic"
 
 	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
+	"golang.org/x/net/context"
 
 	"github.com/cockroachdb/cockroach/pkg/internal/client"
 	"github.com/cockroachdb/cockroach/pkg/kv"
@@ -224,6 +225,11 @@ type distSQLReceiver struct {
 	// Once set, no more rows are accepted.
 	err error
 
+	// cancelled is atomically set to 1 when this distSQL receiver has been marked
+	// as cancelled. Upon the next Push(), err is set to a non-nil
+	// value, and ConsumerClosed is the ConsumerStatus.
+	cancelled int32
+
 	row    parser.Datums
 	status distsqlrun.ConsumerStatus
 	alloc  sqlbase.DatumAlloc
@@ -257,6 +263,7 @@ type rowResultWriter interface {
 }
 
 var _ distsqlrun.RowReceiver = &distSQLReceiver{}
+var _ distsqlrun.CancellableRowReceiver = &distSQLReceiver{}
 
 // makeDistSQLReceiver creates a distSQLReceiver.
 //
@@ -325,6 +332,10 @@ func (r *distSQLReceiver) Push(
 		}
 		return r.status
 	}
+	if r.err == nil && atomic.LoadInt32(&r.cancelled) == 1 {
+		// Set the error to reflect query cancellation.
+		r.err = sqlbase.NewQueryCanceledError()
+	}
 	if r.err != nil {
 		// TODO(andrei): We should drain here.
 		return distsqlrun.ConsumerClosed
@@ -368,6 +379,11 @@ func (r *distSQLReceiver) ProducerDone() {
 		panic("double close")
 	}
 	r.closed = true
+}
+
+// SetCancelled is part of the CancellableRowReceiver interface.
+func (r *distSQLReceiver) SetCancelled() {
+	atomic.StoreInt32(&r.cancelled, 1)
 }
 
 // updateCaches takes information about some ranges that were mis-planned and

--- a/pkg/sql/distsqlrun/base.go
+++ b/pkg/sql/distsqlrun/base.go
@@ -91,6 +91,17 @@ type RowReceiver interface {
 	ProducerDone()
 }
 
+// CancellableRowReceiver is a special type of a RowReceiver that can be set to
+// cancelled asynchronously (i.e. concurrently or after Push()es and ProducerDone()s).
+// Once cancelled, subsequent Push()es return ConsumerClosed. Implemented by distSQLReceiver
+// which is the final RowReceiver, and the origin point for propagation of ConsumerClosed
+// consumer statuses.
+type CancellableRowReceiver interface {
+	// SetCancelled sets this RowReceiver as cancelled. Subsequent Push()es (if any)
+	// return a ConsumerStatus of ConsumerClosed.
+	SetCancelled()
+}
+
 // RowSource is any component of a flow that produces rows that cam be consumed
 // by another component.
 type RowSource interface {

--- a/pkg/sql/distsqlrun/flow.go
+++ b/pkg/sql/distsqlrun/flow.go
@@ -424,9 +424,9 @@ func (f *Flow) RunSync(ctx context.Context) {
 }
 
 // cancel iterates through all unconnected streams of this flow and marks them cancelled.
-// It also closes the syncFlowConsumer, if it exists. This function is called in Wait()
-// after the associated context has been cancelled. In order to cancel a flow, call
-// f.ctxCancel() instead of this function.
+// If the syncFlowConsumer is of type CancellableRowReceiver, mark it as cancelled.
+// This function is called in Wait() after the associated context has been cancelled.
+// In order to cancel a flow, call f.ctxCancel() instead of this function.
 //
 // For a detailed description of the distsql query cancellation mechanism,
 // read docs/RFCS/query_cancellation.md.
@@ -452,11 +452,9 @@ func (f *Flow) cancel() {
 	}
 
 	if f.syncFlowConsumer != nil {
-		// Push an error to the sync flow consumer. If this is a distSQLReceiver,
-		// it will call ConsumerClosed when it sees a non-nil error.
-		f.syncFlowConsumer.Push(
-			nil, /* row */
-			ProducerMetadata{Err: sqlbase.NewQueryCanceledError()})
+		if recv, ok := f.syncFlowConsumer.(CancellableRowReceiver); ok {
+			recv.SetCancelled()
+		}
 	}
 }
 

